docs/Security_Enhancement_For_AI_Models.md

```diff
@@ -0,0 +1,246 @@
+- Register key to Key Broker Service (KBS), and KBS communicates with key management service (KMS) to store the key.
+
+
+### 1.2 Confidential Cloud Native Primitives (CCNP)
+
+[CCNP](https://github.com/intel/confidential-cloud-native-primitives)  is the solution targeted on simplifying the use of Trusted Execution Environment (TEE) in cloud-native environment. There are 2 parts included in CCNP, the services and the SDK.
+
+- Service is designed to hide the complexity of different TEE platforms and provides common interfaces and scalability for cloud-native environment.
+
+- SDK is to simplify the use of the service interface for development, it covers communication to the service and parses the results from the services.
+
+The service supports attestation, measurement fetching and event log collecting of various platforms including Intel Trusted Domain Extensions (TDX), Trusted Platform Modules (TPM) and AMD SEV-SNP.
+
+To get the key to decrypt the model, we need provide the quote of TEE for attestation, CCNP is a good choice to get the quote and it hides the complexity and is easy to use:
+
+```Python
+from ccnp import Quote
+quote=Quote.get_quote()
+```
+
+Detail documentation can be found [here](https://intel.github.io/confidential-cloud-native-primitives/).
+
+### 1.3 Inference Flow
+
+The inference service flow of geting the model should be updated since the model has been encyrpted:
+
+- Deploy CCNP as a [DaemonSet](https://intel.github.io/confidential-cloud-native-primitives/_rst/service.quote.html#deploy-as-daemonset-in-kubernetes), it will detect the TEE environment.
+- Get the quote and request the key from KBS.
+- Get the encrypted model from model server.
+- Decrypt the model and load the model do inference.
+
+## 2. Implementation
+
+Current Cloud Native AI Pipeline (CNAP) already implemented the new design to protect the models, this section will describe the details of each components and introduce how to enable a new model server or KBS interface.
+
+To protect the model, the first thing is generating a key then encypting the model and register the key to KBS, KBS could submit request to store the key to KMS.
+To get a model and decrypt the model, the runtime environment should be verified by attestation. A key broker client should get a quote and send the quote to KBS to attest the environment is TEE then response a wrapped key for decryption.
+
+### 2.1 Mode Encryption
+
+AES GCM is a widely used encryption algorithm and mode of operation for securing data, particularly in network communications and data storage. It combines two important cryptographic techniques: AES and GCM. AES is a symmetric encryption algorithm that uses a shared secret key for both encryption and decryption. GCM is based on a combination of CTR and GMAC, this combination ensures that not only is the data encrypted, but it also includes an authentication tag that allows the recipient to verify the integrity of the data.
```

Comment:
AES-GCM is a widely used block cipher mode of operation ....
And regarding the introduction of GCM, i remember it is an algorithm that runs on a block cipher with block size 128 bits(mainly AES). And GMAC is the authentication-only variant of GCM. Please take the NIST publication as reference. https://nvlpubs.nist.gov/nistpubs/legacy/sp/nistspecialpublication800-38d.pdf

Reply:
GMAC is the authentication-only variant of GCM, and GCM is based on a combination of CTR and GMAC, this should be ok, right?